qiskit/quantum_info/synthesis/qsd.py

```diff
@@ -95,11 +97,23 @@ def decomp_2q(mat):
                 decomposer_2q = two_qubit_decompose.two_qubit_cnot_decompose
         circ = decomposer_2q(mat)
     else:
+        if _is_block_diagonal(mat, dim_o2):
+            # this about halves CX count over standard qsd for this special case
+            u1, u2 = mat[:dim_o2, :dim_o2], mat[dim_o2:, dim_o2:]
+            return _demultiplex(u1, u2, opt_a1=opt_a1, opt_a2=opt_a2, _depth=_depth)
```

Comment:
This is just idle musings that I don't want to do in this PR (because I want to get this in the merge queue ASAP), but if we have a more efficient format for block-diagonal decomposition, is it worth looking into doing a sort of qubit-aware Cuthill-McKee step to reorder the `qargs` of the matrix, and make it more likely that any given matrix is block-diagonal with a lower bandwidth?  If so, we could do the decomposition on a tensor-space permuted matrix, then make sure we apply the output gates to the correct unpermuted qubits?

Reply:
In retrospect, I realise that that's what your "TODO" comment is about inside `_is_block_diagonal`.

Reply:
Yes, I think that would be great. I saw this discussion on [mathoverflow](https://mathoverflow.net/questions/68041/showing-block-diagonal-structure-of-matrix-by-reordering) which mentioned that approach and a couple of others.